src/engine/Preprocessor.cpp

```diff
@@ -781,6 +794,278 @@ void Preprocessor::dumpAllBounds( const String &message )
+    };
+
+    Map<Index, NeuronInformation> indexToNeuron;
+    Map<unsigned, Index> variableToIndex;
+    Map<unsigned, unsigned> layerToLayerSize;
+
+    ASSERT( !_preprocessed._networkLevelReasoner );
+
+    const List<Equation> &equations( _preprocessed.getEquations() );
+    const List<PiecewiseLinearConstraint *> plConstraints( _preprocessed.getPiecewiseLinearConstraints() );
+
+    // If the activation functions are not yet supported by the NLR,
+    // abort
+    for ( const auto constraint : plConstraints )
+        if ( !NetworkLevelReasoner::functionTypeSupported( constraint->getType() ) )
+            return false;
+
+    // Attempt to figure out the layered structure
+    List<unsigned> inputLayer;
+    List<unsigned> previousLayer;
+    List<unsigned> currentLayer;
+    List<unsigned> outputLayer;
+    unsigned currentLayerIndex = 1;
+
+    // The input and output layer should be defined in the input query
+    inputLayer = _preprocessed.getInputVariables();
+    previousLayer = inputLayer;
+    unsigned numVariablesHandledSoFar = inputLayer.size();
+    outputLayer = _preprocessed.getOutputVariables();
+    layerToLayerSize[0] = inputLayer.size();
+
+    // Store the input neurons in the topology
+    unsigned count = 0;
+    for ( const auto &neuron : inputLayer )
+    {
+        NeuronInformation neuronInfo;
+        neuronInfo._activationVariable = neuron;
+        Index index( 0, count );
+        indexToNeuron[index] = neuronInfo;
+        variableToIndex[neuron] = index;
+        ++count;
+    }
+
+    // Now, attempt to figure out the topology, layer by layer
+    while ( true )
+    {
+        currentLayer.clear();
+        for ( const auto &eq : equations )
+        {
+            // Only consider equalities
+            if ( eq._type != Equation::EQ )
+                continue;
+
+            Set<unsigned> eqVars = eq.getParticipatingVariables();
+            for ( const auto &var : previousLayer )
+                eqVars.erase( var );
+
+            if ( eqVars.size() == 1 )
+            {
+                // We have identified the weithed sum for a new neuron
+                NeuronInformation neuronInfo;
+                neuronInfo._weightedSumVariable = *eqVars.begin();
+                neuronInfo._weightedSumEquation = &eq;
+
+                Index index( currentLayerIndex, currentLayer.size() );
+                indexToNeuron[index] = neuronInfo;
+                currentLayer.append( *eqVars.begin() );
+
+                variableToIndex[*eqVars.begin()] = index;
+            }
+        }
+
+        // If a new layer has not been discovered, we're done
+        if ( currentLayer.empty() )
+            break;
+
+        numVariablesHandledSoFar += currentLayer.size();
+        layerToLayerSize[currentLayerIndex] = currentLayer.size();
+
+        // If this layer is the output layer, we're done
+        bool outputLayerFound = false;
+        if ( currentLayer.size() == outputLayer.size() )
```

Comment:
Is this a safe condition to identify output layer?

Reply:
I think cycling through the variables is slightly better, because it accounts for a common scenario where there are additional variables that are not included in the topology. For example, some delta variable that is used for an adversarial robustness bound.

Reply:
Ah, I see.